In [1]:
#importing necessary modules
import pandas as pd
from sklearn.model_selection import train_test_split
from imblearn.over_sampling import SMOTE
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score, classification_report, confusion_matrix
from xgboost import XGBClassifier 
import warnings
warnings.filterwarnings('ignore')

In [2]:

data = pd.read_excel('processed_data.xls')

In [3]:
data.head()

Age  Gender  EducationBackground  MaritalStatus  EmpDepartment  EmpJobRole  \
0   32       1                    2              2              5          13   
1   47       1                    2              2              5          13   
2   40       1                    1              1              5          13   
3   41       1                    0              0              3           8   
4   60       1                    2              2              5          13   

   BusinessTravelFrequency  DistanceFromHome  EmpEducationLevel  \
0                        2                10                  3   
1                        2                14                  4   
2                        1                 5                  4   
3                        2                10                  4   
4                        2                16                  4   

   EmpEnvironmentSatisfaction  ...  EmpRelationshipSatisfaction  \
0                           4  ...                            4   
1                           4  ...                            4   
2                           4  ...                            3   
3                           2  ...                            2   
4                           1  ...                            4   

   TotalWorkExperienceInYears  TrainingTimesLastYear  EmpWorkLifeBalance  \
0                          10                      2                   2   
1                          20                      2                   3   
2                          20                      2                   3   
3                          23                      2                   2   
4                          10                      1                   3   

   ExperienceYearsAtThisCompany  ExperienceYearsInCurrentRole  \
0                            10                             7   
1                             7                             7   
2                            18                            13   
3                            21                             6   
4                             2                             2   

   YearsSinceLastPromotion  YearsWithCurrManager  Attrition  PerformanceRating  
0                        0                     8          0                  3  
1                        1                     7          0                  3  
2                        1                    12          0                  4  
3                       12                     6          0                  3  
4                        2                     2          0                  3  

[5 rows x 27 columns]

In [4]:
data.shape

(1200, 27)

In [5]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1200 entries, 0 to 1199
Data columns (total 27 columns):
 #   Column                        Non-Null Count  Dtype
---  ------                        --------------  -----
 0   Age                           1200 non-null   int64
 1   Gender                        1200 non-null   int64
 2   EducationBackground           1200 non-null   int64
 3   MaritalStatus                 1200 non-null   int64
 4   EmpDepartment                 1200 non-null   int64
 5   EmpJobRole                    1200 non-null   int64
 6   BusinessTravelFrequency       1200 non-null   int64
 7   DistanceFromHome              1200 non-null   int64
 8   EmpEducationLevel             1200 non-null   int64
 9   EmpEnvironmentSatisfaction    1200 non-null   int64
 10  EmpHourlyRate                 1200 non-null   int64
 11  EmpJobInvolvement             1200 non-null   int64
 12  EmpJobLevel                   1200 non-null   int64
 13  EmpJobSatisfaction            120

In [6]:
#deviding the data into predictors and outcome variables
X=data.iloc[:,:-1]
y=data['PerformanceRating']

In [7]:
#splitting the X and y into train and test samples with 20% holdout sample for testing
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split( X, y, test_size=0.25, random_state=42)

In [8]:
#checking out the dimentions of training and testing samples
X_train.shape,X_test.shape,y_train.shape,y_test.shape

((900, 26), (300, 26), (900,), (300,))

In [12]:
#sm = SMOTE()
#X_train, y_train = sm.fit_resample(X_train, y_train)

In [13]:
#scaling the predictor train and test data before training 
sc = StandardScaler()

sc_x_train = sc.fit_transform(X_train) 
sc_x_test = sc.transform(X_test)

# 3.A trained model which can predict the employee performance based on factors as inputs. This will be used to hire employees.


### a) Logistic Regression

In [14]:
#importing the logistic regression classification,training and predicting for test samples 
from sklearn.linear_model import LogisticRegression
model1=LogisticRegression()
model1_fit=model1.fit(X_train,y_train)
y_predict1=model1_fit.predict(X_test)

In [15]:
#importing classification metrics to evaluate the model
from sklearn.metrics import accuracy_score, classification_report,confusion_matrix
print('Accuracy Score:',accuracy_score(y_test,y_predict1))
print()
print('Confusion matrix:\n',confusion_matrix(y_test,y_predict1))
print()
print('Classification Report:\n',classification_report(y_test,y_predict1))

Accuracy Score: 0.7966666666666666

Confusion matrix:
 [[ 19  17   1]
 [ 18 208   6]
 [  0  19  12]]

Classification Report:
               precision    recall  f1-score   support

           2       0.51      0.51      0.51        37
           3       0.85      0.90      0.87       232
           4       0.63      0.39      0.48        31

    accuracy                           0.80       300
   macro avg       0.67      0.60      0.62       300
weighted avg       0.79      0.80      0.79       300



### b) Decision Tree

In [16]:
#importing the Decision tree classification,training and predicting for test samples 
from sklearn.tree import DecisionTreeClassifier
model2=DecisionTreeClassifier()
model2_fit=model2.fit(X_train,y_train)
y_predict2=model2_fit.predict(X_test)

In [17]:
print('Accuracy Score:',accuracy_score(y_test,y_predict2))
print()
print('Confusion matrix:\n',confusion_matrix(y_test,y_predict2))
print()
print('Classification Report:\n',classification_report(y_test,y_predict2))

Accuracy Score: 0.8633333333333333

Confusion matrix:
 [[ 29   8   0]
 [ 12 207  13]
 [  1   7  23]]

Classification Report:
               precision    recall  f1-score   support

           2       0.69      0.78      0.73        37
           3       0.93      0.89      0.91       232
           4       0.64      0.74      0.69        31

    accuracy                           0.86       300
   macro avg       0.75      0.81      0.78       300
weighted avg       0.87      0.86      0.87       300



### c) Random Forest

In [18]:
#importing the Random Forest classification,training and predicting for test samples 

from sklearn.ensemble import RandomForestClassifier
model3=RandomForestClassifier()
model3_fit=model3.fit(X_train,y_train)
y_predict3=model3_fit.predict(X_test)

In [19]:
print('Accuracy Score:',accuracy_score(y_test,y_predict3))
print()
print('Confusion matrix:\n',confusion_matrix(y_test,y_predict3))
print()
print('Classification Report:\n',classification_report(y_test,y_predict3))

Accuracy Score: 0.93

Confusion matrix:
 [[ 33   4   0]
 [  7 224   1]
 [  0   9  22]]

Classification Report:
               precision    recall  f1-score   support

           2       0.82      0.89      0.86        37
           3       0.95      0.97      0.96       232
           4       0.96      0.71      0.81        31

    accuracy                           0.93       300
   macro avg       0.91      0.86      0.88       300
weighted avg       0.93      0.93      0.93       300



### XGBoost Classifier

In [20]:
# Training the model
from xgboost import XGBClassifier
model_xgb = XGBClassifier()
model_xgb.fit(X_train,y_train)

[17:36:05] WARNING: C:/Users/Administrator/workspace/xgboost-win64_release_1.3.0/src/learner.cc:1061: Starting in XGBoost 1.3.0, the default evaluation metric used with the objective 'multi:softprob' was changed from 'merror' to 'mlogloss'. Explicitly set eval_metric if you'd like to restore the old behavior.


XGBClassifier(base_score=0.5, booster='gbtree', colsample_bylevel=1,
              colsample_bynode=1, colsample_bytree=1, gamma=0, gpu_id=-1,
              importance_type='gain', interaction_constraints='',
              learning_rate=0.300000012, max_delta_step=0, max_depth=6,
              min_child_weight=1, missing=nan, monotone_constraints='()',
              n_estimators=100, n_jobs=8, num_parallel_tree=1,
              objective='multi:softprob', random_state=0, reg_alpha=0,
              reg_lambda=1, scale_pos_weight=None, subsample=1,
              tree_method='exact', validate_parameters=1, verbosity=None)

In [21]:
# Predicting the model
y_predict_xgb = model_xgb.predict(X_test)

In [22]:

# Finding accuracy, precision, recall and confusion matrix
print(accuracy_score(y_test,y_predict_xgb))
print(classification_report(y_test,y_predict_xgb))

0.92
              precision    recall  f1-score   support

           2       0.80      0.86      0.83        37
           3       0.94      0.95      0.95       232
           4       0.88      0.74      0.81        31

    accuracy                           0.92       300
   macro avg       0.88      0.85      0.86       300
weighted avg       0.92      0.92      0.92       300



Model with the results(accuracy)

-Logistic Regression      - 79.66 %

-Decision Tree            - 86.33 %

-Random forest classifier - 93 %

-XG Boost                 - 92 %

## 4. Recommendations to improve the employee performance based on insights from analysis.

From the analysis we found out that,
The Variables like Employee EnvironmentSatisfaction,
EmpLastSalaryHikePercent,
EmpWorkLifeBalance,
ExperienceYearsInCurrentRole, 
OverTime are more related to the performance ratings, So company need to focus on

We can conclude that the company should provide a better environment as it increases the performance drastically. The company should increase the salary of the employee from time to time and help them maintain a worklife balance.

Conclusion
For this project, we used algorithms like Logistic Regression, Decision Tree, Random Forest to calculate the accuracy
and found out that Random Forest gives the maximum accuracy of 93%.